### PR TITLE
HPCC-18561 WUDetails Regression script

### PR DESCRIPTION
**Pull request by shamser**
Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by shamser** [comment]
@AttilaVamos and I have been discussing this branch.  Let's continue discussion here so that we can have feedback from @ghalliday and @afishbeck too.

**Comment by AttilaVamos** [comment]
From our discussion:
I start to play with your ESP test code and I have some problem, thoughts:
your code uses an external (not standard) package zeep, which can be problematic. I installed it via pip and I got:
```
pip list --format=columns | grep zeep
zeep                         2.5.0    
```
but your code doesn't work with it:
```
./wutest.py 
Traceback (most recent call last):
  File "./wutest.py", line 26, in <module>
    from zeep import Client 
ImportError: No module named 'zeep'
```
Other question is:
- Does your implementation works with SSL enable esp server?

For your question: "Does ecl command have an option to check the wsdl?"
I don't know, but if it doesn't, you can implement one and it has several advantages, e.g no external dependencies, if there is a problem it will be solved in house and quickly, experts review you code and help you, it produces exactly what you need, you don't need to worrying about SSL and LDAP authentication, etc.

**Comment by shamser** [comment]
The test script is called wutest.py.  The usage instructions may be found in README.txt.  All the other files are the expected results for each test case (within the key sub-directory)

**Comment by AttilaVamos** [comment]
This is a huge changes: 212 files, +2,733,118 lines, ~108 MB. Actually smoketest system went to meditative state during get the diff file. So there is not smoketest result for this PR until I fix it.

**Comment by shamser** [comment]
The script is python3.  So, try this: pip3 list --format=columns |grep zeep

It does support SSL.  If you provide a url with 'https://... ' it will connect through ssl.  By the way if you use self-signed certificates, you'll probably won't to use the --nosslverify option.

The usage instructions are:
 wutest.py [-h] [-o dir] [-d] [--logresp] [--logreq] [-n days]
                 [-t number] [--nosslverify] [-u url] [--user username]
                 [--pw password] [--httpdigestauth]

All arguments are optional.  By default, it will use WUQuery and WUDetails services running on the local server.

As part of the regression, it will query the WSDL for WUDetails and WUQuery.  WSDL changes that could cause the behaviour of WUDetails to be different should generate an error- new fields should be ok but if any existing fields are removed or modified, an error should be generated.

**Comment by shamser** [comment]
@AttilaVamos  Regarding zeep being a non-standard library:  there doesn't appear to be built in support for soap.  (I'm a complete Python nubie so I could be wrong on that point.)  From reading https://wiki.python.org/moin/WebServices#SOAP, it appears that all of the SOAP packages are 'non-standard'.

**Comment by AttilaVamos** [comment]
@shamser My concern is not if you are using Python or Python3 to implement a test system. I do same to Regression Test Engine and Smoketest. My problem is the lack of functionality, support and the dependency of external library. 

That is why I suggested to you to use our ECL and/or SOAPPLUS tools and I showed you a code sample how I used it from Python. If some functionality is missing from ECL or SOAPPLUS tool, then raise a JIRA and implement it (as I did with ECL to add  abort, status, zapgen, etc. commands and plan to do it to get some stats)  or ask an expert to do that for you. 

Of course everything above is only a hint and example how I do my stuff.

If you like to use external library then it is fine, but prepare your code to handle if it isn’t installed and give the user a proper error message and a guide how to install it. And add it into the documentation as well.

**Comment by AttilaVamos** [comment]
Smoketest is fixed, this PR is scheduled to test and it is the 3rd in the queue.

**Comment by ghalliday** [comment]
@ghalliday

**Comment by AttilaVamos** [comment]
Something wrong with this script:
```
./wutest.py 
Traceback (most recent call last):
  File "./wutest.py", line 128, in GetTestWorkunits
    if (('Workunits' not in wuqueryresp) or ('ECLWorkunit' not in wuqueryresp['Workunits'])):
TypeError: argument of type 'NoneType' is not iterable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./wutest.py", line 497, in <module>
    wu = GetTestWorkunits()
  File "./wutest.py", line 156, in GetTestWorkunits
    print('Error unexpected response from WUQuery:', js) 
NameError: name 'js' is not defined
```

**Comment by shamser** [comment]
@AttilaVamos Please can you refresh you copy and try again.

**Comment by AttilaVamos** [comment]
@shamser This is progressed:
```
./wutest.py 
[INFO] Starting new HTTP connection (1): localhost
[INFO] Starting new HTTP connection (1): localhost
[INFO] Starting new HTTP connection (1): 192.168.1.58
```
But I can't see any result neither on the console screen nor in the 'results' sub-directory.

File: testing/esp/wudetails/wutest.py [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+from requests.auth import HTTPBasicAuth
+from requests.auth import HTTPDigestAuth
+from zeep import Client 
+from zeep.transports import Transport
+from zeep.cache import SqliteCache
+from zeep.plugins import HistoryPlugin
+from collections import namedtuple
+from argparse import Namespace
+from pathlib import Path
+from lxml import etree
```
Comment (ghalliday) [inline testing/esp/wudetails/wutest.py]: not actually used.

File: testing/esp/wudetails/wutest.py [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+import requests.packages.urllib3
+from requests import Session
+from requests.auth import HTTPBasicAuth
+from requests.auth import HTTPDigestAuth
+from zeep import Client 
+from zeep.transports import Transport
+from zeep.cache import SqliteCache
+from zeep.plugins import HistoryPlugin
+from collections import namedtuple
+from argparse import Namespace
```
Comment (ghalliday) [inline testing/esp/wudetails/wutest.py]: not used

File: testing/esp/wudetails/wutest.py [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+
+requiredJobs = ('childds1','dedup_all','sort','keydiff','key', 'keyed_join', 'mergechild', 'dict1', 'indexread2')
+
+maskValueFields = ('Definition','SizePeakMemory', 'WhenFirstRow', 'TimeElapsed', 'TimeTotalExecute', 'TimeFirstExecute', 'TimeLocalExecute',
+                   'WhenStarted', 'TimeMinLocalExecute', 'TimeMaxLocalExecute', 'TimeAvgLocalExecute', 'SkewMinLocalExecute', 'SkewMaxLocalExecute',
+                   'NodeMaxLocalExecute', 'NodeMaxDiskWrites', 'NodeMaxLocalExecute', 'NodeMaxLocalExecute', 'NodeMaxSortElapsed', 'NodeMinDiskWrites',
+                   'NodeMinLocalExecute', 'NodeMinLocalExecute', 'NodeMinLocalExecute', 'NodeMinSortElapsed', 'SkewMaxDiskWrites', 'SkewMaxLocalExecute',
+                   'SkewMaxLocalExecute', 'SkewMaxSortElapsed', 'SkewMinDiskWrites', 'SkewMinLocalExecute', 'SkewMinLocalExecute', 'SkewMinSortElapsed',
+                   'TimeAvgSortElapsed', 'TimeMaxSortElapsed', 'TimeMinSortElapsed')
+maskMeasureTypes = ('ts','ns','sz', 'skw', 'node', 'cnt')
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply testing/esp/wudetails/wutest.py]: I think we need to discuss hthor and thor counts on Tuesday - e.g., HPCC-18217.

**Review by ghalliday — COMMENTED** [review]
Generally looks good.  A few comments.

**Comment by ghalliday** [comment]
@shamser see comments above

File: testing/esp/wudetails/wutest.py [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+logging.basicConfig(level=loglevel, format='[%(levelname)s] %(message)s')
+
+keydir = 'key'
+wuqueryservice_url = args.baseurl + '/WsWorkunits/WUQuery.json?ver_=1.71'
+wuqueryservice_wsdl_url = args.baseurl + '/WsWorkunits/WUQuery.json?ver_=1.71&wsdl'
+wudetailsservice_wsdl_url = args.baseurl + '/WsWorkunits/WUDetails.json?ver_=1.71&wsdl'
+
+requiredJobs = ('childds1','dedup_all','sort','keydiff','key', 'keyed_join', 'mergechild', 'dict1', 'indexread2')
+
+maskValueFields = ('Definition','SizePeakMemory', 'WhenFirstRow', 'TimeElapsed', 'TimeTotalExecute', 'TimeFirstExecute', 'TimeLocalExecute',
```
[the remark this answers is not in the document]
Comment (shamser) [reply testing/esp/wudetails/wutest.py]: @ghalliday If the the IncludeMeasure='0' property option is used, then the test below cannot be used as not measure is returned.

File: testing/esp/wudetails/wutest.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shamser) [reply testing/esp/wudetails/wutest.py]: For some work units, cnt and sz field change from run to run.  I'll find out which ones these are.

**Comment by ghalliday** [comment]
@shamser I think this is back to you, but not picked up as assigned to you

File: testing/esp/wudetails/wutest.py (hunk 1 above)
Comment (shamser) [reply testing/esp/wudetails/wutest.py]: Done

File: testing/esp/wudetails/wutest.py [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+argumentparser.add_argument('-u', '--baseurl', help="Base url for both WUQuery and WUDetails", default='http://localhost:8010', metavar='url')
+argumentparser.add_argument('--user', help='Username for authentication', metavar='username')
+argumentparser.add_argument('--pw', help='Password for authentication', metavar='password')
+argumentparser.add_argument('--httpdigestauth', help='User HTTP digest authentication(Basic auth default)', action='store_true')
+args = argumentparser.parse_args()
+
+if (args.debug):
+    loglevel = logging.DEBUG
+else:
+    loglevel=logging.INFO
```
[the remark this answers is not in the document]
Comment (shamser) [reply testing/esp/wudetails/wutest.py]: Done

File: testing/esp/wudetails/wutest.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shamser) [reply testing/esp/wudetails/wutest.py]: Created a new ticket to reduce size of comparison file: https://track.hpccsystems.com/browse/HPCC-19077

**Comment by shamser** [comment]
A Jira ticket has been created to modify Thor to always return the stats that are sometimes missed out: https://track.hpccsystems.com/browse/HPCC-19077

**Comment by shamser** [comment]
@ghalliday Please can you review the changes

File: testing/esp/wudetails/wutest.py [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+print('--------------------------------')
+print('')
+
+argumentparser = argparse.ArgumentParser()
+argumentparser.add_argument('-o', '--outdir', help='Results directory', default='results', metavar='dir')
+argumentparser.add_argument('-d', '--debug', help='Enable debug', action='store_true', default=False)
+argumentparser.add_argument('--logresp', help='Log wudetails responses (in results directory)', action='store_true', default=False)
+argumentparser.add_argument('--logreq', help='Log wudetails requests (in results directory)', action='store_true', default=False)
+argumentparser.add_argument('-n', '--lastndays', help="Use workunits from last 'n' days", type=int, default=1, metavar='days')
+argumentparser.add_argument('-t', '--testcase', help="Execute give testcase", type=int, default=0, metavar='number')
```
Comment (ghalliday) [inline testing/esp/wudetails/wutest.py]: did you mean "given"?

Comment (shamser) [reply testing/esp/wudetails/wutest.py]: I've actually removed this functionality.  The functionality makes less sense given that the executing full execution of all the test cases is quick... updating

**Pull request merged**